### PR TITLE
Staging/jesd fsm JESD204 (FSM) Interface Framework

### PR DESCRIPTION
**Pull request by danmois**
Add the possibility of using the JESD204 (FSM) Interface Framework.

### CODE REVIEW

File: drivers/axi_core/jesd204/axi_jesd204_tx.c [hunk 1]
```diff
@@ -247,7 +256,39 @@ uint32_t axi_jesd204_tx_status_read(struct axi_jesd204_tx *jesd)
...
+	chksum += config->ctrl_bits_per_sample;
+	chksum += config->converter_resolution - 1;
+	chksum += config->bits_per_sample - 1;
+	chksum += config->subclass;
+	chksum += config->samples_per_conv_frame ? config->samples_per_conv_frame - 1 :
+		  0;
+	chksum += config->jesd_version;
+	chksum += config->high_density;
+
+	return chksum & 0xff;
```
Comment (CiprianRegus) [inline drivers/axi_core/jesd204/axi_jesd204_tx.c:282]: `NO_OS_GENMASK()`

File: drivers/axi_core/jesd204/axi_jesd204_tx.c [hunk 2]
```diff
@@ -276,8 +317,53 @@ uint32_t axi_jesd204_tx_calc_ilas_chksum(const struct jesd204_tx_config *config)
...
+			break;
+		case 1:
+			val = lane_id;
+			val |= (config->num_lanes - 1) << 8;
+			val |= config->scrambling << 15;
+			val |= (config->octets_per_frame - 1) << 16;
+			val |= (config->frames_per_multiframe - 1) << 24;
+			break;
+		case 2:
+			val = (config->num_converters - 1);
```
Comment (CiprianRegus) [inline drivers/axi_core/jesd204/axi_jesd204_tx.c:340]: Remove the parentheses.

File: drivers/axi_core/jesd204/axi_jesd204_tx.c [hunk 3]
```diff
@@ -364,12 +510,301 @@ int32_t axi_jesd204_tx_apply_config(struct axi_jesd204_tx *jesd,
...
+static int axi_jesd204_tx_jesd204_link_pre_setup(struct jesd204_dev *jdev,
+		enum jesd204_state_op_reason reason,
+		struct jesd204_link *lnk)
+{
+	struct axi_jesd204_tx_jesd204_priv *priv = jesd204_dev_priv(jdev);
+	struct axi_jesd204_tx *jesd = priv->jesd;
+	unsigned long link_rate, device_rate, lane_rate;
+#if 0
+	long rate;
+#endif
```
Comment (CiprianRegus) [inline drivers/axi_core/jesd204/axi_jesd204_tx.c:535]: Why is this necessary (same goes for all the `#if 0` directives below)?

File: drivers/axi_core/jesd204/axi_jesd204_tx.c [hunk 4]
```diff
@@ -364,12 +510,301 @@ int32_t axi_jesd204_tx_apply_config(struct axi_jesd204_tx *jesd,
...
+{
+	struct axi_jesd204_tx_jesd204_priv *priv = jesd204_dev_priv(jdev);
+	struct axi_jesd204_tx *jesd = priv->jesd;
+
+	pr_debug("%s:%d link_num %u reason %s\n", __func__, __LINE__, lnk->link_id,
+		 jesd204_state_op_reason_str(reason));
+
+	switch (reason) {
+	case JESD204_STATE_OP_REASON_INIT:
+		break;
```
Comment (CiprianRegus) [inline drivers/axi_core/jesd204/axi_jesd204_tx.c:745]: You can return `JESD204_STATE_CHANGE_DONE`.

File: drivers/axi_core/jesd204/axi_jesd204_tx.c [hunk 5]
```diff
@@ -460,6 +895,81 @@ int32_t axi_jesd204_tx_init(struct axi_jesd204_tx **jesd204,
...
+	struct axi_jesd204_tx *jesd;
+	uint32_t synth_1;
+	uint32_t magic;
+	uint32_t version;
+	uint32_t tmp;
+	int ret;
+
+	jesd = (struct axi_jesd204_tx *)malloc(sizeof(*jesd));
+	if (!jesd)
+		return -1;
```
Comment (CiprianRegus) [inline drivers/axi_core/jesd204/axi_jesd204_tx.c:917]: `-ENOMEM` ?

File: drivers/axi_core/jesd204/axi_jesd204_tx.c [hunk 6]
```diff
@@ -460,6 +895,81 @@ int32_t axi_jesd204_tx_init(struct axi_jesd204_tx **jesd204,
...
+{
+	struct axi_jesd204_tx_jesd204_priv *priv;
+	struct axi_jesd204_tx *jesd;
+	uint32_t synth_1;
+	uint32_t magic;
+	uint32_t version;
+	uint32_t tmp;
+	int ret;
+
+	jesd = (struct axi_jesd204_tx *)malloc(sizeof(*jesd));
```
Comment (CiprianRegus) [inline drivers/axi_core/jesd204/axi_jesd204_tx.c:915]: `calloc()`

File: drivers/adc/ad9680/ad9680.c [hunk 7]
```diff
@@ -109,6 +115,114 @@ int32_t ad9680_test(struct ad9680_dev *dev,
...
+	return JESD204_STATE_CHANGE_DONE;
+}
+
+int ad9680_jesd_enable_link(struct ad9680_dev *dev, uint8_t en)
+{
+	int err;
+	uint8_t tmp_reg;
+
+	if (dev == NULL)
+		return -11;
```
Comment (CiprianRegus) [inline drivers/adc/ad9680/ad9680.c:158]: Why are we returning the explicit error code instead of a macro?

File: drivers/adc/ad9680/ad9680.c [hunk 8]
```diff
@@ -109,6 +115,114 @@ int32_t ad9680_test(struct ad9680_dev *dev,
...
+
+	if (en > 1)
+		return -14;
+
+	err = ad9680_spi_read(dev, AD9680_REG_JESD_LINK_CTRL1_REG, &tmp_reg);
+	if (err)
+		return err;
+	tmp_reg &= ~AD9680_JESD_LINK_PDN;
+	tmp_reg |= en ? 0 : AD9680_JESD_LINK_PDN;
+	err = ad9680_spi_write(dev, AD9680_REG_JESD_LINK_CTRL1_REG, tmp_reg);
```
Comment (CiprianRegus) [inline drivers/adc/ad9680/ad9680.c:168]: `return ad9680_spi_write()`

File: drivers/adc/ad9680/ad9680.c (hunk 7 above)
Comment (danmois) [reply drivers/adc/ad9680/ad9680.c:158]: This has been preserved for sync with Linux variant. Replaced by ENOMEM and EINVAL.

**Comment by danmois** [comment]
V2

- implemented review comments

**Comment by danmois** [comment]
V3

- remove unused code from file axi_jesd204_tx.c

File: drivers/axi_core/jesd204/axi_jesd204_rx.c [hunk 9]
```diff
@@ -550,6 +614,314 @@ int32_t axi_jesd204_rx_apply_config(struct axi_jesd204_rx *jesd,
...
+}
+
+static int axi_jesd204_rx_jesd204_link_pre_setup(struct jesd204_dev *jdev,
+		enum jesd204_state_op_reason reason,
+		struct jesd204_link *lnk)
+{
+	struct axi_jesd204_rx_jesd204_priv *priv = jesd204_dev_priv(jdev);
+	struct axi_jesd204_rx *jesd = priv->jesd;
+	unsigned long link_rate, lane_rate, device_rate;
+#if 0
```
[the remark this answers is not in the document]
Comment (danmois) [reply drivers/axi_core/jesd204/axi_jesd204_rx.c:634]: I will remove these lines. The implementation will follow in a subsequent PR.

**Comment by danmois** [comment]
V4

- remove unused code from file axi_jesd204_rx.c
- implement review comments

File: projects/fmcdaq2/src/app/fmcdaq2.c [hunk 10]
```diff
@@ -990,9 +990,9 @@ static int fmcdaq2_setup(struct fmcdaq2_dev *dev,
 	if (status != 0)
 		return status;
 
-	status = ad9144_setup(&dev->ad9144_device, &dev_init->ad9144_param);
+	status = ad9144_setup_legacy(&dev->ad9144_device, &dev_init->ad9144_param);
```
Comment (buha) [inline projects/fmcdaq2/src/app/fmcdaq2.c:1140]: why a legacy ? aren't we supposed to use the new jesd setup ?

File: projects/fmcdaq2/src/app/fmcdaq2.c [hunk 11]
```diff
@@ -462,6 +533,15 @@ static int fmcdaq2_trasnceiver_setup(struct fmcdaq2_dev *dev,
 {
 	int status;
 
+#ifdef JESD_FSM_ON
```
Comment (buha) [inline projects/fmcdaq2/src/app/fmcdaq2.c:536]: same as my other comment, i think we should use the jesd framework and not care about the old setup method

File: projects/fmcdaq2/src/app/fmcdaq2.c (hunk 10 above)
Comment (danmois) [reply projects/fmcdaq2/src/app/fmcdaq2.c:1140]: This function was renamed to "_legacy" because the new JESD FSM implementation uses a new setup function with the same name . This renamed one is the old setup function.

File: projects/fmcdaq2/src/app/fmcdaq2.c (hunk 11 above)
Comment (danmois) [reply projects/fmcdaq2/src/app/fmcdaq2.c:536]: For the time being we will keep both implementations (with and without using the FSM).

**Review by buha — APPROVED** [review]
i got nothing more to add, but don't forget with @dbogdan those unimplemented/empty ops

**Review by dbogdan — APPROVED** [review]

**Review by buha — APPROVED** [review]
